File: web/src/types/timetable.ts
/** Athon — Timetable Type Definitions */

export interface PeriodInfo {
  id: string;
  name: string;
  period_number: number;
  start_time: string;
  end_time: string;
  is_break: boolean;
}

export interface TimetableTeacherInfo {
  id: string;
  name: string;
  employee_code: string;
}

export interface TimetableSubjectInfo {
  id: string;
  name: string;
  code: string;
  is_core: boolean;
}

export interface TimetableClassInfo {
  id: string;
  name: string;
  section: string | null;
}

export interface TimetableEntry {
  id: string;
  day_of_week: number;
  room_number: string | null;
  is_active: boolean;
  period: PeriodInfo;
  subject: TimetableSubjectInfo;
  teacher: TimetableTeacherInfo;
  class_: TimetableClassInfo;
}

export interface TimetableResponse {
  entries: TimetableEntry[];
}

export interface CreateTimetableEntryRequest {
  teacher_id: string;
  class_id: string;
  subject_id: string;
  period_id: string;
  academic_term_id: string;
  day_of_week: number;
  room_number?: string;
}

export interface UpdateTimetableEntryRequest {
  teacher_id?: string;
  class_id?: string;
  subject_id?: string;
  period_id?: string;
  day_of_week?: number;
  room_number?: string;
  is_active?: boolean;
}

export interface ClassTimetableOption {
  id: string;
  name: string;
  section: string | null;
}

File: backend/app/api/schemas/timetable.py
"""Pydantic schemas for timetable API endpoints."""

from datetime import time

from pydantic import BaseModel


class PeriodInfo(BaseModel):
    """Brief period info included in timetable responses."""

    id: str
    name: str
    period_number: int
    start_time: time
    end_time: time
    is_break: bool


class SubjectInfo(BaseModel):
    """Brief subject info included in timetable responses."""

    id: str
    name: str
    code: str
    is_core: bool


class TeacherInfo(BaseModel):
    """Brief teacher info included in timetable responses."""

    id: str
    name: str
    employee_code: str


class ClassInfo(BaseModel):
    """Brief class info included in timetable responses."""

    id: str
    name: str
    section: str | None = None


class TimetableEntryResponse(BaseModel):
    """A single timetable entry with resolved related entity info.

    All related objects are pre-loaded to prevent N+1 queries.
    """

    id: str
    day_of_week: int
    room_number: str | None = None
    is_active: bool

    # Resolved related entities
    period: PeriodInfo
    subject: SubjectInfo
    teacher: TeacherInfo
    class_: ClassInfo

    model_config = {"from_attributes": True}


class TimetableResponse(BaseModel):
    """Full timetable response — either a weekly or daily schedule."""

    entries: list[TimetableEntryResponse]
